Add changelog

mgrauer opened the pull request:
@girder/developers 

I've added a proposed changelog, including the release notes for 2.3.0.

If people like this, we can copy the notes from here to the release itself in GitHub, and people can start adding to the CHANGELOG as significant pull requests are merged.  This should make future release note generation much simpler and hopefully will also push us to have more frequent releases.

I'd love to take some of the overall maintenance burden off of @zachmullen and spread it around the community, this is a step towards that.

If people agree to this, I'd want to wait for @brianhelba 's 2.3.0 client side additions and any other release note additions or clarifications before merging.

jcfr commented:
Well done.

Optionally, if you would like to easily add issue and user references. You could make use of https://github.com/sloria/sphinx-issues#usage

danlamanna commented:
A few notes:

- Whether or not we use the plugin @jcfr mentioned, I think relevant PRs
should be added to each item (going forward). We do a pretty good job
cross-referencing issues and PRs so users should be able to work their way
backwards if they have to.
- We should settle on rst or md, so far we're a mixture (README is rst,
CONTRIBUING is md). rst probably makes the most sense since it's the format
of our docs.
- We might want to add a note on semantic versioning and how/where we
follow it.


On Sat, Jul 29, 2017 at 4:43 PM, Jean-Christophe Fillion-Robin <
notifications@github.com> wrote:

> Well done.
>
> Optionally, if you would like to easily add issue and user references. You
> could make use of https://github.com/sloria/sphinx-issues#usage
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/girder/girder/pull/2232#issuecomment-318858055>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAlH5cC3JarsnevykdXRJJ58Ty1PoBcMks5sS5l9gaJpZM4OncxK>
> .
>



-- 
Dan LaManna
Kitware, Inc.

danlamanna commented:
Also I think #2040 is big enough to warrant mentioning in the added section.

On Mon, Jul 31, 2017 at 12:31 PM, Dan LaManna <notifications@github.com>
wrote:

> A few notes:
>
> - Whether or not we use the plugin @jcfr mentioned, I think relevant PRs
> should be added to each item (going forward). We do a pretty good job
> cross-referencing issues and PRs so users should be able to work their way
> backwards if they have to.
> - We should settle on rst or md, so far we're a mixture (README is rst,
> CONTRIBUING is md). rst probably makes the most sense since it's the format
> of our docs.
> - We might want to add a note on semantic versioning and how/where we
> follow it.
>
>
> On Sat, Jul 29, 2017 at 4:43 PM, Jean-Christophe Fillion-Robin <
> notifications@github.com> wrote:
>
> > Well done.
> >
> > Optionally, if you would like to easily add issue and user references.
> You
> > could make use of https://github.com/sloria/sphinx-issues#usage
> >
> > —
> > You are receiving this because you are on a team that was mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/girder/girder/pull/2232#issuecomment-318858055>, or
> mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AAlH5cC3JarsnevykdXRJJ58Ty1PoBcMks5sS5l9gaJpZM4OncxK>
> > .
> >
>
>
>
> --
> Dan LaManna
> Kitware, Inc.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/girder/girder/pull/2232#issuecomment-319124797>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAlH5Y6tUa60qifpt908qhjP0tS9Ct5Iks5sTgF4gaJpZM4OncxK>
> .
>



-- 
Dan LaManna
Kitware, Inc.

zachmullen commented:
> If people like this, we can copy the notes from here to the release itself in GitHub, and people can start adding to the CHANGELOG as significant pull requests are merged. This should make future release note generation much simpler and hopefully will also push us to have more frequent releases.

The one thing I feel kind of strongly about is that we should keep only one authoritative source for the changelog, it should not be duplicated between this file and the github release notes. So I would propose that we do one of the following:

1. Have this changelog capture every single version, and don't put any release notes in the GH releases.
2. Have this changelog just contain the unreleased changes, and when a new release is packaged, we dump its contents into GH release notes and empty them as part of the release process.

The second option would be my preference, only because we already do keep our changelog on the GH releases, and it meshes nicely into that UI.

danlamanna commented:
The first option goes against the grain but I think it's a better option
since it captures the entirety of the changelog inside the actual repo. In
that case our release notes can just be a link to the
changelog#release-heading.

On Mon, Jul 31, 2017 at 1:54 PM, Zach Mullen <notifications@github.com>
wrote:

> If people like this, we can copy the notes from here to the release itself
> in GitHub, and people can start adding to the CHANGELOG as significant pull
> requests are merged. This should make future release note generation much
> simpler and hopefully will also push us to have more frequent releases.
>
> The one thing I feel kind of strongly about is that we should keep only
> one authoritative source for the changelog, it should not be duplicated
> between this file and the github release notes. So I would propose that we
> do one of the following:
>
>    1. Have this changelog capture every single version, and don't put any
>    release notes in the GH releases.
>    2. Have this changelog just contain the unreleased changes, and when a
>    new release is packaged, we dump its contents into GH release notes and
>    empty them as part of the release process.
>
> The second option would be my preference, only because we already do keep
> our changelog on the GH releases, and it meshes nicely into that UI.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/girder/girder/pull/2232#issuecomment-319145437>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAlH5dXimLVLOBZI2ky73kZ023mQpgkiks5sThTigaJpZM4OncxK>
> .
>



-- 
Dan LaManna
Kitware, Inc.

zachmullen commented:
@danlamanna that's fine by me. As long as it's not duplicated in two places, I'm happy. In fact, if I don't have to edit GH release notes, I'm even happier.

jcfr commented:
@zachmullen  Just to clarify, the change logs will end up being duplicated (github repo, readthedocs, pypi, ... ). What will not be duplicated is the effort to generate it for multiple places.

Adding a link back to readthedocs from the GitHub release will definitively streamline the release process.

brianhelba commented:
@mgrauer Your local instance of Git does not have its `user.name` properly set, such that your commits are attributed to just `mgrauer` in Git. I'd suggest you fix this and re-write the branch.

mgrauer commented:
Also #2229, which makes me think twice about linking individual PRs.  Sometimes features are added in bunches of PRs with a number of incremental additions or bug fixes, that would make more sense to be summarized as an overall feature rather than broken out into individual PRs.

I'll get back to this later in the week, I have been travelling and will be again tomorrow.

🕴

mgrauer commented:
I rewrote my commits and added everything I think needs to be there for 2.3.0.  PTAL.

I'm sure I'm missing other updates post 2.3.0, please add them or post a comment informing me and I can update--of course these can be always be added on a later PR, the important thing for now is anything before the 2.3.0 release.

zachmullen commented:
Let's not let the perfect be the enemy of the good, let's go ahead and merge it. Or to put it tersely,

#FQAMTS

mgrauer commented:
I think that is what he was getting at, that is, FQAMTS essentially recapitulates "the perfect is the enemy of the good".

To be fair, this changelog has had a bit of QA.

mgrauer commented:
I have added a link from the GitHub release to the 2.3.0 section of the CHANGELOG file.

It's not clear to me what further process people want, please create issues for any additional needs, including

* describing how we do semantic versioning
* linking between the readthedocs and CHANGELOG
* linking to PRs/Issues/Users from the CHANGELOG
* use the Github release tool
* use the Sphinx issues tool

jcfr commented:
Thanks Mike. Looks great.

> linking between the readthedocs and CHANGELOG

Since CHANGELOG is a rst file, this is already addressed i think

mgrauer commented:
@jcfr Are you saying readthedocs will automatically add our CHANGELOG?  Wouldn't it have to be added into the structure of our docs and in the docs folder for that to happen?

jcfr commented:
> into the structure of our docs and in the docs folder for that to happen

Yes. I though that was the case.

It is quite easy. Here is an example: https://github.com/scikit-build/scikit-build/blob/master/docs/changes.rst
